zkvm-host/bin/agg.rs

```diff
@@ -0,0 +1,123 @@
+    let workspace_root = metadata.workspace_root;
+    let proof_directory = format!("{}/data/proofs", workspace_root);
+
+    let mut proofs = Vec::with_capacity(proof_names.len());
+    let mut boot_infos = Vec::with_capacity(proof_names.len());
+
+    for proof_name in proof_names.iter() {
+        let proof_path = format!("{}/{}.bin", proof_directory, proof_name);
+        if fs::metadata(&proof_path).is_err() {
+            panic!("Proof file not found: {}", proof_path);
+        }
+        let mut deserialized_proof =
+            SP1ProofWithPublicValues::load(proof_path).expect("loading proof failed");
+        proofs.push(deserialized_proof.proof);
+
+        // The public values are the ABI-encoded BootInfo.
+        let mut boot_info_buf = [0u8; BOOT_INFO_SIZE];
+        deserialized_proof
+            .public_values
+            .read_slice(&mut boot_info_buf);
+        let boot_info = RawBootInfo::abi_decode(&boot_info_buf).unwrap();
+        boot_infos.push(boot_info);
+    }
+
+    (proofs, boot_infos)
+}
+
+// Execute the Kona program for a single block.
+#[tokio::main]
+async fn main() -> Result<()> {
+    utils::setup_logger();
+
+    let args = Args::parse();
+    let fetcher = SP1KonaDataFetcher::new();
+    let prover = ProverClient::new();
+
+    let (proofs, boot_infos) = load_aggregation_proof_data(args.proofs);
+
+    // Fetch the headers from the L1 head of the first boot info to the current L1 head, inclusive.
+    let first_head = boot_infos.last().unwrap().l1_head;
+
+    // Confirm that the headers are in the correct order.
+    let start_header = fetcher
+        .get_header_by_hash(ChainMode::L1, first_head)
+        .await?;
+    // End header is the current L1 head.
+    // TODO: Should this be the finalized header/fetched from a different source?
+    let end_header = fetcher.get_head(ChainMode::L1).await?;
+    if start_header.number > end_header.number {
+        panic!("Headers are not in the correct order");
+    }
+    let mut headers = Vec::new();
+
+    let mut curr_header = end_header.clone();
+    // Fetch the headers from the end header to the start header, inclusive.
+    while curr_header.number >= start_header.number {
+        headers.push(curr_header.clone());
+        curr_header = fetcher
+            .get_header_by_hash(ChainMode::L1, curr_header.parent_hash)
+            .await?;
+    }
+
+    // Reverse the headers to put them in order from start to end.
+    headers.reverse();
+
+    let (_, vkey) = prover.setup(MULTI_BLOCK_ELF);
+
+    let mut stdin = SP1Stdin::new();
+    for proof in proofs {
+        let SP1Proof::Compressed(compressed_proof) = proof else {
+            panic!();
+        };
+        stdin.write_proof(compressed_proof, vkey.vk.clone());
+    }
+
+    // Write the aggregation inputs to the stdin.
+    stdin.write(&AggregationInputs {
+        boot_infos,
+        headers,
+        l1_head: end_header.hash_slow(),
```

Comment:
Once we pass the final L1 Head in as an arg, we can just use that instead of rehashing.